### ISSUE TITLE
Proposal: ref-argument version of IComparer and/or Comparison

### ISSUE THREAD

**Issue by kg** [dotnet/corefx#33927]
For scenarios where you want to sort an array or list of structs, the overhead of pass-by-value for IComparer/Comparison can become quite significant, especially if the structs are large. In my experience it's difficult to ensure all the comparisons get inlined by the runtime (though I'm sure there's always room for improvement there) - the ideal solution (and one I've implemented by hand) is to define an alternative IRefComparer<T> interface that takes (ref T lhs, ref T rhs) instead of (T lhs, T rhs). It's very easy to define an adapter to go between IComparer and IRefComparer when necessary (in my case my sorter uses IRefComparer with an adapter that passes the ref'd arguments directly to an IComparer, when needed).

For my main test scenario (sorting a large number of draw calls / batches before sending them to the GPU), switching from IComparer to IRefComparer (using my customized version of the BCL sort routine) drops total CPU usage for the render prepare in my profiles from 10% to 7% - so I saved 3% off of my total CPU usage just by going from by-value (which generates lots of pushes and pops for the draw call structs) to by-ref.

The existing overloads that take a keys+values pair of arrays and sort the keys are a partial solution here, because the sorting operation doesn't have to swap the (huge) values around when sorting - but the values are still passed to the comparator by-value, so you get at least two copies per comparison. It adds up.

My basic proposal here is that a new optional interface is introduced - IRefComparer<T> or IComparerRef<T> - and a basic implementation of it is provided for all the BCL types that have an IComparer available. This might mean IComparableRef would also need to be introduced, I'm not quite sure how all the default types fit together in the BCL? From my perspective only IComparer needs it since that is the type used by most list/array sort primitives. Then with the interface introduced, I'd probably just suggest adding overloads to Array.Sort and List.Sort, with those being the two main APIs people would need to sort through (other than LINQ, I suppose?)

One problem with this proposal I don't know how to solve is that you could end up with overload ambiguity between IComparer and IRefComparer (or between IComparable and IRefComparable, if those exist). I think the only solution there would be to require the Ref version of the interface to inherit the non-Ref version, so that it's possible to use the ref version as the preferred type and just pass it to APIs that don't support ref comparisons.

**Comment by dzmitry-lahoda** [comment]
```
Array.SortInPlace<MyRefSorter>(arrayOrSpan) where MyRefSorter : IRefComparer, struct 
{
  MyRefSorter sorter = default;
  ...
}
``
Some kind of constrained generic with inlining? to be SAME fast as manual? No delegates (5% of perf and array version 5% perf).

**Comment by kg** [comment]
> Some kind of constrained generic with inlining?

Yeah, that's roughly how it looks. The performance improvement is dramatic even on classic netframework, and I bet more aggressive inlining on ref + devirtualization on the generics would be massive.

Having any help from the BCL or runtime to simplify it would be pretty cool - the one inconvenience is having to wrap the ref comparer for APIs that only take a regular one. But you can always implement both interfaces with your comparer

**Comment by dzmitry-lahoda** [comment]
Related proposals for sorting. I guess sorting should use some kind of comparer. And comparer could be by `ref`(actually `in`).

https://github.com/dotnet/corefx/issues/15329
https://github.com/EamonNerbonne/anoprsst/tree/master/src/Anoprsst
https://github.com/DotNetCross/Sorting

**Comment by kg** [comment]
Definitely seems like if you're sorting Spans you shouldn't be passing the lhs and rhs by value, yeah.
